design/migration-path.md

```diff
@@ -0,0 +1,92 @@
+## Proposal
+
+### Conan recipe compatible syntax 1.X => 2.0
+
+This doesn't mean that all Conan 1.X syntax and features will work in Conan 2.0. Many things will be deprecated and removed.
+It means that Conan will provide the features that will be defined as the canonical and recommended ones in Conan 2.0 in 1.X.
+These features will be declared stable in 1.X when there is enough usage evidence, to allow companies to safely adopt them.
+As an example (and this shouldn't be discussed in this proposal, it is just an example), the current ``CMake`` build helper will be
+deprecated, and using the new ``CMakeToolchain()`` plus the new build helper that uses the generated toolchain file will be the
+feature that will work in Conan 1.X and will continue to work in Conan 2.0.
+
+Implementing this forward compatibility is critical for the upgrade. Conan 2.0 cannot require in recipes any new feature
+that is not already supported in Conan 1.X. Users, (and that includes **ConanCenter** central repo) will need to gradually adapt
+and prepare their recipes during 1.X, adopting the features that will keep working in Conan 2.0.
+
+### Conan command line will change in 2.0
+
+This is the only possibility. There is a lot of evidence that some commands are not the most intuitive (e.g. the ``search`` should
+by default search in remotes, not in the cache. Again, just an example, not something to discuss here), and the only possible way
+forward is to change the command line syntax and behavior. Contrary to the Conan recipes, here it is not possible to implement
+a "compatible forward" transition, and users will need to upgrade their scripts using Conan to the new Conan 2.0 syntax.
+
+Some new commands have already started to be designed with some new guidelines for the Conan 2.0 CLI. For example, the ``conan lock``
+command uses always the full path to the files, like ``conan lock create conanfile.py``, because the Conan 2.0 will be more explicit,
+to avoid some confusing ambiguities of the current CLI syntax. But the other commands will need to change, and in most cases, it is
+impossible to do it in a non-breaking way, for example ``conan install path/conanfile.py`` would be incorrectly processed as a ``pkg_name/version``.
+
+### Behavior will change in 2.0
+
+The fact that recipes will be forward compatible doesn't imply that the behavior will be the same. Some things might change,
+for example (again, as an example, not to be discussed here), the default ``package_id_mode`` might change, and then it might become necessary
+building new binaries. Or the ``build_requires`` recipes will always be fetched, even if they are not needed for building from source,
+making the installs a bit less efficient.
+
+There are a few Conan defaults that made sense a few years ago, but not anymore nowadays, like the default ``settings.libcxx=libstdc++``, even
+for gcc>5 compilers. Conan will change this and use ``libstdc++11`` whenever available, but this is breaking behavior.
```

Comment:
I'm in favor of the overall document and I can see most of the transition going pretty smoothly (with the measures outlined in the document and the PR comments). The only thing that I'd really want to watch out for is this section: behavior changes for identical recipes. This is because it has the potential for silent breaks. If the recipe syntax changes, it's going to explicitly fail and that's obvious. However, if the same recipe does something different after the version change, that's harder to detect.

I can see these kinds of changes being necessary but it's definitely the main aspect I'll be watching out for. The things that would be good from my point of view:
* Document these clearly, including a transition guide.
* When possible, ensure that the new behavior is stricter than the old one. For example, going to a stricter default `package_id_mode` may cause some inefficiencies during the build process but it will not cause any ABI issues. That's fine. The other way around, if the new behavior became less strict, would result in silent ABI problems for anyone who missed that point in the transition docs. (My example here may be a bit contrived. Based on the doc, I assume you're already going in a stricter direction.)